### PR TITLE
Update DFA to consider ReDim as null-safe validator

### LINKED ISSUE #3437: CA1062 - Get error assigning a value to an array element
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (latest release)

#### Diagnostic ID

Example: [CA1062](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1062?view=vs-2019)

#### Repro steps
`    Public Sub GetValues(ByRef Values() As String)`
`          ReDim Values(0)`
`          Values(0) = "Test Value" ' <-- Error on array element`
`    End Sub`

#### Expected behavior
No error due to CA1062 being violated

The array when passed is basically a pointer to the start of the array and the ReDim will erase it and reinitialize it.

Most likely the first element in the array at this point would be a null value, but assigning something to it would at that point make it not null.

#### Actual behavior
I get a CA1062 error on the line indicated above.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #3437

@Youssef1313 I was doing some experiment to try to help you out and doing so I actually did the implementation. As you can see there wasn't much to add.

### CODE REVIEW

**Comment by mavasani** [comment]
I believe a better fix to handle this similar to `VisitObjectCreation` as following:

1. Override this Visit method in the base `DataFlowOperationVisitior` type with appropriate default value returned by that visitor
2. Override this Visit method in `PointsToDataFlowOperationVisitor` and `AbstractLocationDataFlowOperationVisitor`. Implementation would likely be similar to `VisitObjectCreation` overrides in these visitors.

This will ensure ReDim is handle correctly by all the DFA based analyses instead of just CA1062. I think if you do the above, you will likely not need any changes to ParameterValidationAnalysis visitor.

**Comment by Evangelink** [comment]
@mavasani I have tried to apply your suggestions but I am not sure I did it correctly. I also think I would need to add more tests to ensure the new behavior is correct for the other locations I changed but honestly I haven't really understood all the changes so I'll need some help/input.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 1]
```diff
@@ -763,6 +764,23 @@ public override PointsToAbstractValue VisitAnonymousObjectCreation(IAnonymousObj
                 return VisitTypeCreationWithArgumentsAndInitializer(operation, argument, base.VisitAnonymousObjectCreation);
             }
 
+            public override PointsToAbstractValue VisitReDimClause(IReDimClauseOperation operation, object? argument)
+            {
+                var type = operation.Children.FirstOrDefault()?.Type;
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs:769]: Please do not use `Children` API

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs [hunk 2]
```diff
@@ -1260,7 +1260,8 @@ private void PerformPredicateAnalysis(IOperation operation)
                 operation.Kind == OperationKind.Invocation ||
                 operation.Kind == OperationKind.Argument ||
                 operation.Kind == OperationKind.FlowCaptureReference ||
-                operation.Kind == OperationKind.IsPattern);
+                operation.Kind == OperationKind.IsPattern ||
+                operation.Kind == OperationKind.ReDimClause);
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:1264]: This should not be added here. ReDim clause can never be a predicate returning boolean value.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs [hunk 3]
```diff
@@ -3361,6 +3362,16 @@ public override TAbstractAnalysisValue VisitAwait(IAwaitOperation operation, obj
                 value;
         }
 
+        public override TAbstractAnalysisValue VisitReDimClause(IReDimClauseOperation operation, object? argument)
+        {
+            var value = base.VisitReDimClause(operation, argument);
+            if (PredicateAnalysis)
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:3368]: ReDim clause can never be a predicate, this is not correct.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs (hunk 1 above)
Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs:769]: Did you mean to just use the Operand?

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 4]
```diff
@@ -763,6 +764,23 @@ public override PointsToAbstractValue VisitAnonymousObjectCreation(IAnonymousObj
...
+                    return base.VisitReDimClause(operation, argument);
+                }
+
+                AbstractLocation location = AbstractLocation.CreateAllocationLocation(operation, type, DataFlowAnalysisContext);
+                var pointsToAbstractValue = PointsToAbstractValue.Create(location, mayBeNull: false);
+                CacheAbstractValue(operation, pointsToAbstractValue);
+
+                _ = base.VisitReDimClause(operation, argument);
+
+                return pointsToAbstractValue;
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs:781]:
```suggestion
                return VisitTypeCreationWithArgumentsAndInitializer(operation, argument, base.VisitReDimClause);
```

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 5]
```diff
@@ -763,6 +764,23 @@ public override PointsToAbstractValue VisitAnonymousObjectCreation(IAnonymousObj
                 return VisitTypeCreationWithArgumentsAndInitializer(operation, argument, base.VisitAnonymousObjectCreation);
             }
 
+            public override PointsToAbstractValue VisitReDimClause(IReDimClauseOperation operation, object? argument)
+            {
+                var type = operation.Children.FirstOrDefault()?.Type;
+                if (type == null)
+                {
+                    return base.VisitReDimClause(operation, argument);
+                }
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs:773]:
```suggestion
                if (operation.Operand.Type == null)
                {
                    return base.VisitReDimClause(operation, argument);
                }
```

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs (hunk 2 above)
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:1264]:
```suggestion
                operation.Kind == OperationKind.IsPattern);
```

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs (hunk 3 above)
Comment (Evangelink) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:3368]: Ok so what's the point of overriding the method in this class if the only thing is to return the base?

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs (hunk 4 above)
Comment (Evangelink) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs:781]: This is causing 2 debug assert to fail, that's why I did this edited copy-paste.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs (hunk 1 above)
Comment (Evangelink) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs:769]: Yes!

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs [hunk 6]
```diff
@@ -3361,6 +3362,16 @@ public override TAbstractAnalysisValue VisitAwait(IAwaitOperation operation, obj
                 value;
         }
 
+        public override TAbstractAnalysisValue VisitReDimClause(IReDimClauseOperation operation, object? argument)
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs]: @mavasani Have you been able to give some thoughts on this?

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.ParameterValidationDataFlowOperationVisitor.cs [hunk 7]
```diff
@@ -225,6 +225,13 @@ public override ParameterValidationAbstractValue Visit(IOperation operation, obj
                 return value;
             }
 
+            public override ParameterValidationAbstractValue VisitReDimClause(IReDimClauseOperation operation, object? argument)
+            {
+                var value = base.VisitReDimClause(operation, argument);
+                MarkValidatedLocations(operation.Operand);
+                return value;
+            }
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.ParameterValidationDataFlowOperationVisitor.cs:233]: @Evangelink Can you try if reverting the `VisitReDimClause` override in this file has any effect? I think your changes to the next two files might make this override redundant. If not, then feel free to retain this override.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs]: Your changes seems fine to me now.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.ParameterValidationDataFlowOperationVisitor.cs (hunk 7 above)
Comment (Evangelink) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.ParameterValidationDataFlowOperationVisitor.cs:233]: Tests `ReDim_FirstInstruction_NoDiagnostic` and `ReDim_FirstInstructionMultipleVariables_NoDiagnostic` are failing (they report the diagnostic while they shouldn't) if I revert this change.

Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.ParameterValidationDataFlowOperationVisitor.cs:233]: No worries - your change LGTM. Thanks!

**Pull request merged**